### ISSUE TITLE
ESLint 6.2.2 + babel-eslint 10.0.3 + no-unused-vars false positive with for-of loop

### ISSUE THREAD

**Issue by camyyssa** [eslint/eslint#12187]
**Tell us about your environment**

* **ESLint Version: 6.2.2**
* **Node Version: 12.6.0**
* **npm Version: 6.11.2**

**What parser (default, Babel-ESLint, etc.) are you using?**
Babel-ESLint `10.0.3`

Also tried Babel-ESLint `10.0.2` but ran into https://github.com/eslint/eslint/issues/12117. Following the advice there, I tried running `10.0.3`

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  parser: 'babel-eslint',
  parserOptions: {
    ecmaVersion: 6,
    sourceType: 'module',
    ecmaFeatures: {
      jsx: true,
      generators: true,
      experimentalObjectRestSpread: true
    }
  },
  extends: ["airbnb"],
  rules: {
    'react/sort-comp': 0,
    "react/require-default-props": 0,
    "react/forbid-prop-types": ['error', { forbid: ['any', 'array']}],
    "react/destructuring-assignment": 0,
    "react/static-property-placement": ['error', 'static public field'],
    "react/state-in-constructor": 0,
    "react/jsx-props-no-spreading": 0,

    "linebreak-style": 0,
    "function-paren-newline": ['error', 'consistent'],
    "object-curly-newline": ['error', { 'consistent': true }],

    "babel/semi": 2,
    "no-param-reassign": ['error', { props: false }],
    "arrow-parens": ['error', 'as-needed', { "requireForBlockBody": true }],
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
for (const type of TYPES) {
  const itemsOfType = items.filter(item => item.__typename === type);
}
```

```js
      for (const [name, offset] of items) {
        if (offset >= scrollTop) {
          this.setState({ active: name });
          return;
        }
      }
```

```bash
npm run lint
```

**What did you expect to happen?**
Get no errors

**What actually happened? Please include the actual, raw output from ESLint.**
```
  98:10  error  'type' is defined but never used             no-unused-vars
  126:15  error  'name' is defined but never used          no-unused-vars
  126:21  error  'offset' is defined but never used          no-unused-vars
```

**Are you willing to submit a pull request to fix this bug?**
Yes, but I can't commit to it in the next couple of weeks

**Comment by davidroeca** [comment]
Encountering the same

**Comment by platinumazure** [comment]
Could anyone encountering this issue please try running `npm ls eslint-scope` (or equivalent) and show the output?

**Comment by sombhotlu** [comment]
`-- react-scripts@3.1.1
  +-- @typescript-eslint/eslint-plugin@1.13.0
  | `-- @typescript-eslint/experimental-utils@1.13.0
  |   `-- eslint-scope@4.0.3  deduped
  +-- babel-eslint@10.0.2
  | `-- eslint-scope@3.7.1 
  `-- webpack@4.39.1
    `-- eslint-scope@4.0.3

**Comment by sombhotlu** [comment]
I was facing same issue when using updated `react-script@3.1.1`  which suggests `babel-eslint@10.0.2`, but then I manually updated to `babel-eslint@10.0.3` it and all the warnings went away.

**Comment by camyyssa** [comment]
@platinumazure output from `yarn list eslint-scope` on my project: 

```
yarn list v1.17.3
warning Filtering by arguments is deprecated. Please use the pattern option instead.
├─ babel-eslint@10.0.3
│  └─ eslint-scope@3.7.1
├─ eslint-scope@4.0.3
└─ eslint@6.2.2
   └─ eslint-scope@5.0.0
✨  Done in 1.21s.
```

Regarding the comment from @sombhotlu, I have tried updating to `babel-eslint@10.0.3` (it was already mentioned in #12117 ). The original `no-unused-vars` errors disappeared, but in my project they were replaced by `no-unused-vars` in different locations ¯\_(ツ)_/¯ Thank you for mentioning a possible solution though! It might help other readers.

**Comment by platinumazure** [comment]
@camyyssa It looks like your babel-eslint is still using the old eslint-scope for some reason. Try blowing away the node_modules and rerunning `npm install`, does that resolve your issue?

**Comment by camyyssa** [comment]
😳 It does!

I could have found the steps to resolve this on my own 😊. Sorry to waste your time @platinumazure  🙂 but thank you so much for the help! 

For future me, and in case you stumble on this post because you have the same problem: 
1. Upgrade dependency, remove `node-modules`, and try again
2. If you're using the library as a sub-dependency, try to fix it to the latest version using yarn resolutions https://yarnpkg.com/lang/en/docs/selective-version-resolutions/ (doesn't seem to exist on npm)
> This is a temporary solution, to check if updating the dependency solves the problem. It looks dangerous to me to keep this in your `package.json` long-term, since it forces the package manager to install a certain version of a sub-dependency
